### PR TITLE
added audio players

### PR DESCRIPTION
**Pull request by derricklee43**
## Summary of Changes
 - Added hook to play and pause card audio
   - The number of times the definition is repeated is customizable 
   - Exposes the card key and active side of the card being played 
    - Returns a promise 
      - In the future, we can return the results of user response 
 - Added Skeleton of play lesson hook
   -  Controls which cards are played and the order of cards
- Added basic study page

Don't mind the audio, it is pretty off. 
Also, the page no longer shows text of active card audio, instead it is 'front' or 'back'
https://user-images.githubusercontent.com/39753553/187047113-cb9b3363-bb46-4318-96e4-9f04e025cc4b.mov

### CODE REVIEW

File: src/pages/study-page/study-page.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    fetchDeckAndRefresh();
+  }, [deckId]);
+
+  if (deck === undefined) {
+    return <LoadingPage label="loading deck..." />;
+  }
+
+  return (
+    <div className="study-page">
+      <PageHeader label={deck.metaData.title} onGoBackClick={noop} goBackLabel="Go back" />
```
Comment (jiejasonliu) [inline src/pages/study-page/study-page.tsx:30]: lowercase `go back` to be consistent with other pages

File: src/hooks/use-play-card-audio.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+  function playAudio(audio: HTMLAudioElement) {
+    setActiveAudio(audio);
+    chainToAudio(() => setActiveAudio(undefined), audio);
+    audio.play();
+  }
+
+  return [
+    activeCardKey,
+    activeCardSide,
```
[the remark this answers is not in the document]
Comment (derricklee43) [reply src/hooks/use-play-card-audio.ts]: Yeah, I like that. It's fixed now.

File: src/hooks/use-timer.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import { useRef } from 'react';
+
+export function useTimer() {
```
[the remark this answers is not in the document]
Comment (derricklee43) [reply src/hooks/use-timer.ts:3]: Yeah, that is a good idea. I added some tests. Were there any other tests in particular that you would like me to add?

File: src/hooks/use-timer.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      !timerIdRef.current ||
+      !remainingTimeRef.current ||
+      !startTimeRef.current
+    ) {
+      return;
+    }
+
+    window.clearTimeout(timerIdRef.current);
+    timerIdRef.current = undefined;
+    remainingTimeRef.current -= Date.now() - startTimeRef.current;
```
[the remark this answers is not in the document]
Comment (derricklee43) [reply src/hooks/use-timer.ts]: Good catch. I added a max function so a negative value would resume at 0.

[the remark this answers is not in the document]
Comment (jiejasonliu) [reply src/hooks/use-timer.ts]: your current change isn't quite right, `Date.now() - startTimeRef` will always be positive.

it's when you subtract that value from the `remainingTimeRef` that can be negative

File: src/hooks/use-timer.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jiejasonliu) [reply src/hooks/use-timer.ts:3]: nope! thanks for doing this, just wanted to add confidence this hook works so we don't have some weird bug later 😃

File: src/hooks/use-timer.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jiejasonliu) [reply src/hooks/use-timer.ts]: something else that came up, what happens if i pause the player twice in a row?

let's say the audio is 5s long,
play for 2s -> pause (3s remaining) -> wait for 2 seconds -> resume and pause again

doesn't the logic result in `remainingTimeLeft` to be 1s despite only playing for 2s?

File: src/hooks/test/use-timer.test.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
+import { renderHook, act } from '@testing-library/react-hooks';
```
Comment (jiejasonliu) [inline src/hooks/test/use-timer.test.ts:1]: remove one of these tests, either from test/ or tests/

File: src/hooks/use-timer.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (derricklee43) [reply src/hooks/use-timer.ts]: haha, good catch again. I hopefully have it fixed for real this time, lol

[the remark this answers is not in the document]
Comment (derricklee43) [reply src/hooks/use-timer.ts]: I added another test doing what your comment described and the test seems to work and not lose time while paused. 

When it gets unpaused the startingTime gets updated, or is there another reason why you think the timer loses time while paused?

[the remark this answers is not in the document]
Comment (jiejasonliu) [reply src/hooks/use-timer.ts]: whoops, it was an observation, i had no idea. 

i think your guard case prevents this behavior so thanks for writing a test to ensure that it works.

**Pull request merged**